src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -211,7 +220,67 @@ IPushMessages BuildMessagePump()
             return messagePumpFactory();
         }
 
-        ReceiveConfiguration configuration;
+        static void LoadMessageHandlers(Configuration configuration, List<Type> orderedTypes, IConfigureComponents container, IList<Type> availableTypes)
+        {
+            var types = new List<Type>(availableTypes);
+
+            foreach (var t in orderedTypes)
+            {
+                types.Remove(t);
+            }
+
+            types.InsertRange(0, orderedTypes);
+
+            ConfigureMessageHandlersIn(configuration, types, container);
+        }
+
+        static void ConfigureMessageHandlersIn(Configuration configuration, IEnumerable<Type> types, IConfigureComponents container)
+        {
+            var handlerRegistry = configuration.MessageHandlerRegistry ?? new MessageHandlerRegistry();
+
+            foreach (var t in types.Where(IsMessageHandler))
+            {
+                container.ConfigureComponent(t, DependencyLifecycle.InstancePerUnitOfWork);
+                handlerRegistry.RegisterHandler(t);
+            }
+
+            container.RegisterSingleton(handlerRegistry);
```

Comment:
I know we need to do this for backwards compatibility, but shouldn't we be able to remove our internal usage of resolving the registry from the container with these changes? e.g. in line 70 in this class.

Reply:
I thought we take a staged approach. Similar to the scanned types. First, convert to a component. And then replace usage of types collection with the hosting component. But I'm all ears.